examples/vision/pointnet.py

```diff
@@ -0,0 +1,295 @@
+        return self.activation(x)
+
+
+"""
+PointNet consists of two core components. The primary MLP network, and the transformer
+net (t-net). The t-net aims to learn an affine transformation matrix by its own mini
+network. The mini network strongly resembles the main network. The t-net is used twice.
+The first time to transform the input features ($n \times 3$) into a canonical
+representation. The second is an affine transformation for alignment in feature space ($n
+\times 64$). As per the original paper we constrain the transformation to be close to an
+orthogonal matrix (i.e. $||X*X^T - I||$ = 0).
+"""
+
+
+class OrthogonalRegularizer(keras.regularizers.Regularizer):
+    def __init__(self, k, l2reg=0.001):
+        self.k = k
+        self.l2reg = l2reg
+        self.eye = K.eye(self.k)
+
+    def __call__(self, x):
+        x = K.reshape(x, (-1, self.k, self.k))
+        xxt = K.batch_dot(x, x, axes=(2, 2))
+        return K.sum(self.l2reg * K.square(xxt - self.eye))
+
+
+"""
+ We can define the t-net as a functional model.
+"""
+
+
+def tnet(n_pts, k, name):
+
+    inputs = keras.Input(shape=(n_pts, k))
+
+    # Initalise bias as the indentity matrix
+    bias = keras.initializers.Constant(np.eye(k).flatten())
+    reg = OrthogonalRegularizer(k)
+
+    x = ConvBN(64)(inputs)
+    x = ConvBN(128)(x)
+    x = ConvBN(1024)(x)
+
+    x = layers.GlobalMaxPooling1D()(x)
+
+    x = DenseBN(512)(x)
+    x = DenseBN(256)(x)
+    x = layers.Dense(
+        k * k,
+        kernel_initializer="zeros",
+        bias_initializer=bias,
+        activity_regularizer=reg,
+    )(x)
+
+    feat_T = layers.Reshape((k, k))(x)
+
+    # Apply affine transformation to input features
+    outputs = layers.Dot(axes=(2, 1))([inputs, feat_T])
+
+    return keras.Model(inputs=inputs, outputs=outputs, name=name)
+
+
+"""
+The main network can be then implemented in the same manner where the t-net mini models
+can be dropped in a layers in the graph. Here we replicate the network architecture
+published in the original paper.
+"""
+
+inputs = keras.Input(shape=(N_PTS, 3))
+
+# x = tnet(N_PTS, 3, 'input_tnet')(inputs)
+
+x = ConvBN(64)(inputs)
+x = ConvBN(64)(x)
+
+# x = tnet(N_PTS, 64, 'feat_tnet')(x)
+
+x = ConvBN(64)(x)
+x = ConvBN(128)(x)
+x = ConvBN(1024)(x)
+
+x = layers.GlobalMaxPooling1D()(x)
+
+x = DenseBN(512)(x)
+x = layers.Dropout(0.3)(x)
+
+x = DenseBN(256)(x)
+x = layers.Dropout(0.3)(x)
+
+outputs = layers.Dense(NUM_CLASSES, activation="softmax")(x)
+
+model = keras.Model(inputs=inputs, outputs=outputs, name="pointnet")
+
+model.summary()
+
+"""
+### Train model
+"""
+
+"""
+Once the model is defined it can be trained like any other standard classification model
+using `.compile()` and `.fit()`.
+"""
+
+model.compile(
+    loss="sparse_categorical_crossentropy",
+    optimizer=keras.optimizers.Adam(learning_rate=0.001),
+    metrics=["sparse_categorical_accuracy"],
+)
+
+model.fit(train_dataset, epochs=10, validation_data=test_dataset)
```

Comment:
After training, it would help to visualize some predictions. How long does training take on a K80?

Reply:
Validation accuracy usually hit around 80-90% after 10 epochs, which is inline with expected performance. 1 epoch = 30s on google colab gpu (k80?), so 5 minutes.